### PR TITLE
Ibft Height Manager

### PR DESCRIPTION
**Pull request by rain-on**
The IbftHeightManager is responsible for all things 'meta-round'
related, such things as:
* Handling RoundTimeout and starting a new round
* Handling NewRound messages and starting a new round
* Ensuring RoundChange messages are sent at the correct time with
  appropriate content.
* Collating RoundChange messages and starting a new round using the
  best prepared certificate in the collection.

### CODE REVIEW

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java [hunk 1]
```diff
@@ -12,35 +12,217 @@
...
+    LOG.info("Received a commit message.");
+    actionOrBufferMessage(
+        msg,
+        () -> currentRound.handleCommitMessage(msg),
+        (roundstate) -> roundstate.addCommitMessage(msg));
+  }
+
+  private void actionOrBufferMessage(
+      final SignedData<? extends Payload> msg,
+      final Action inRoundHandler,
```
Comment (jframe) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java]: Be nicer to have stronger typing here for the inRoundHandler function reference so say something like `Consumer<SignedData<T>> inRoundHandler` where T is `<T extends Payload>`

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java [hunk 2]
```diff
@@ -12,35 +12,217 @@
...
+      final Consumer<RoundState> buffer) {
+    final ConsensusRoundIdentifier msgRoundId = msg.getPayload().getRoundIdentifier();
+    if (msgRoundId.equals(currentRound.getRoundIdentifier())) {
+      inRoundHandler.apply();
+    } else if (msgRoundId.getRoundNumber() > currentRound.getRoundIdentifier().getRoundNumber()) {
```
Comment (jframe) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java]: perhaps extract the msgRound.getRoundNumber() > currentRound.getRoundIdentifier().getRoundNumber as this a common check that being used in many places e.g. in handleRoundChangeMessage and handleNewRoundMessage

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java [hunk 3]
```diff
@@ -12,35 +12,217 @@
...
+    }
+  }
 
-  public void handleNewRoundMessage(final SignedData<NewRoundPayload> newRoundMsg) {}
+  private void startNewRound(final int roundNumber) {
+    if (futureRoundStateBuffer.containsKey(roundNumber)) {
+      currentRound =
+          roundFactory.createNewRoundWithState(
+              parentHeader, futureRoundStateBuffer.get(roundNumber));
+      futureRoundStateBuffer.entrySet().removeIf(k -> k.getKey() <= roundNumber);
```
Comment (jframe) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java]: nit: if you use the futureRoundStateBuffer.keySet you can slightly simply the lambda as you won't need to call getKey

Comment (rain-on) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java]: done.

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java (hunk 2 above)
Comment (rain-on) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java]: Done - sort of, introduced the concept of "MessageAge".

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java (hunk 1 above)
Comment (rain-on) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java]: That will work, the goal here was to have the lambdas fully defined in the calling function - i.e. this function no longer actually needs the msg (other than for its RoundNumber) - as the supplied lambda already has the message contained within it.

Have replaced 'Action' with the inbuilt Runnable.

By having the lambda contain the message, this function is no longer templated, which is a secondary nicety.

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java [hunk 4]
```diff
@@ -12,35 +12,241 @@
...
+
+  public void handleNewRoundMessage(final SignedData<NewRoundPayload> msg) {
+    MessageAge messageAge = determineAgeOfMessage(msg);
+    if (messageAge == MessageAge.PRIOR_ROUND) {
+      LOG.info("Received NewRound Message for a prior round.");
+      return;
+    }
+
+    if (newRoundMessageValidator.validateNewRoundMessage(msg)) {
+
```
Comment (Errorific) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java]: extra whitespace line

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java [hunk 5]
```diff
@@ -12,35 +12,241 @@
...
+    LOG.info("Received a commit message.");
+    actionOrBufferMessage(
+        msg,
+        () -> currentRound.handleCommitMessage(msg),
+        (roundstate) -> roundstate.addCommitMessage(msg));
+  }
+
+  private void actionOrBufferMessage(
+      final SignedData<? extends Payload> msg,
+      final Runnable inRoundHandler,
```
Comment (Errorific) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java]: Don't love Runnable as it's documented to be for actions run in a thread. I can't find a suitable alternative in java.util.function that matches the type signature. Do we know of anything?

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java [hunk 6]
```diff
@@ -137,14 +137,14 @@ private RoundChangeStatus storeRoundChangeMessage(final SignedData<RoundChangePa
    *
    * @param completedRoundIdentifier round identifier that has been identified as superseded
    */
-  public void discardCompletedRound(final ConsensusRoundIdentifier completedRoundIdentifier) {
+  public void discardRoundsPriorTo(final ConsensusRoundIdentifier completedRoundIdentifier) {
     roundChangeCache
         .entrySet()
-        .removeIf(entry -> isAnEarlierOrEqualRound(entry.getKey(), completedRoundIdentifier));
+        .removeIf(entry -> isAnEarlierRound(entry.getKey(), completedRoundIdentifier));
```
Comment (Errorific) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: not part of your change, but can simplify with `.keySet().removeIf(roundIdentifier-> isAnEarlierRound(roundIdentifier, ...`

File: consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+    }
+
+    buildCreatedBlock();
+
+    final MessageValidator messageValidator = mock(MessageValidator.class);
+    when(messageValidator.addSignedProposalPayload(any())).thenReturn(true);
+    when(messageValidator.validateCommmitMessage(any())).thenReturn(true);
+    when(messageValidator.validatePrepareMessage(any())).thenReturn(true);
+
+    // when(finalState.getValidators()).thenReturn(validators);
```
Comment (Errorific) [inline consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManagerTest.java]: comment?

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/IbftStateMachine.java [hunk 8]
```diff
@@ -27,7 +27,7 @@ public IbftStateMachine(final IbftBlockCreatorFactory blockCreatorFactory) {
   /**
    * Attempt to consume the event and update the maintained state
    *
-   * @param event the external action that has occurred
+   * @param event the external Action that has occurred
```
Comment (Errorific) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/IbftStateMachine.java]: This got capitalised in an auto refactor from the Action type?

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java [hunk 9]
```diff
@@ -147,36 +151,31 @@ public void roundExpired(final RoundExpiry expire) {
...
+        msg, currentRound::handleCommitMessage, (roundstate) -> roundstate.addCommitMessage(msg));
   }
 
-  private void actionOrBufferMessage(
-      final SignedData<? extends Payload> msg,
-      final Runnable inRoundHandler,
+  private <T extends Payload> void actionOrBufferMessage(
+      final SignedData<T> msg,
+      final Consumer<SignedData<T>> inRoundHandler,
       final Consumer<RoundState> buffer) {
```
Comment (Errorific) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java]: with inRoundHandler accepting the message but the buffer lambda having the msg already inside it, feels mismatched. Need to make them both take msg or neither. I'm leaning towards both and making buffer be BiConsumer<RoundState, SignedData<T>>

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java (hunk 6 above)
Comment (rain-on) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: Done.

File: consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManagerTest.java (hunk 7 above)
Comment (rain-on) [reply consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManagerTest.java]: removed.

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/IbftStateMachine.java (hunk 8 above)
Comment (rain-on) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/IbftStateMachine.java]: done.

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java (hunk 9 above)
Comment (rain-on) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java]: Done.

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java (hunk 5 above)
Comment (rain-on) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java]: I'd argue the docs are providing purpose without appreciating the concept of something which can be "done" - i.e. its a consumer without parameter, and its a supplier without a result.
Anyway - its been changed to a Biconsumer.

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java (hunk 4 above)
Comment (rain-on) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java]: done.

**Pull request merged**